alibi_detect/utils/pytorch/kernels.py

```diff
@@ -20,21 +56,307 @@ def sigma_median(x: torch.Tensor, y: torch.Tensor, dist: torch.Tensor) -> torch.
 
     Returns
     -------
-    The computed bandwidth, `sigma`.
+    The logrithm of the computed bandwidth, `log-sigma`.
     """
     n = min(x.shape[0], y.shape[0])
     n = n if (x[:n] == y[:n]).all() and x.shape == y.shape else 0
     n_median = n + (np.prod(dist.shape) - n) // 2 - 1
     sigma = (.5 * dist.flatten().sort().values[n_median].unsqueeze(dim=-1)) ** .5
-    return sigma
+    return sigma.log()
 
 
-class GaussianRBF(nn.Module):
+class KernelParameter:
     def __init__(
         self,
-        sigma: Optional[torch.Tensor] = None,
-        init_sigma_fn: Callable = sigma_median,
-        trainable: bool = False
+        value: torch.Tensor = None,
+        init_fn: Optional[Callable] = None,
+        requires_grad: bool = False,
+        requires_init: bool = False
+    ) -> None:
+        """
+        Parameter class for kernels.
+
+        Parameters
+        ----------
+        value
+            The pre-specified value of the parameter.
+        init_fn
+            The function used to initialize the parameter.
+        requires_grad
+            Whether the parameter requires gradient.
+        requires_init
+            Whether the parameter requires initialization.
+        """
+        super().__init__()
+        self.value = nn.Parameter(value if value is not None else torch.ones(1),
+                                  requires_grad=requires_grad)
+        self.init_fn = init_fn
+        self.requires_init = requires_init
+
+
+class BaseKernel(nn.Module):
```

Comment:
For sum/product/... kernels, can we abstract the operation away roughly as follows (with an extension to `torch.Tensor` + `__add__` etc)? Or is it not worth abstracting this away?

```python
class CombinedKernel(nn.Module):
   def __init__(self, kernels: nn.ModuleList, operation: str = 'sum') -> None:
      super().__init__()
      self.kernels = kernels
      self.operation = get_operation(operation)

   def forward(self, x: torch.Tensor, y: torch.Tensor, infer_parameters: bool = False) -> torch.Tensor:
      return torch.stack([k(x, y, infer_parameters) for k in self.kernels]).self.operation(dim=0)

   def __add__(self, other):
    
   ...

class SumKernel(CombinedKernel):
   def __init__(self, kernels: nn.ModuleList):
      super().__init__(kernels, 'sum')
```

Reply:
Comment based on offline discussion:

I believe with the current design, performing elementary operations on `BaseKernel`'s, for example the following:

```python
Kernel_avg = (Kernel_0 + Kernel_1) / torch.tensor(2.0)
```

will return something that is no longer ` BaseKernel` (`Kernel_avg` will be a `SumKernel` I think?). This means lots of the code in places such as `MMDDrift` is no longer quite accurate (type hints, `isinstance` checks etc), since we are expecting a `BaseKernel`. It might be worth rethinking the top-level kernel classes (e.g. maybe `SingleKernel` and `CombinedKernel`, which both inherit from a very generic `BaseKernel`). This might also make serialization easier.

Reply:
Regarding serialization, we need to decide how much configurability we want in the `config.toml`. Assuming we have:

```python
Kernel_avg = (Kernel_0 + Kernel_1) / torch.tensor(2.0)
```

the first option is to attempt to serialize the individual kernels separately, and save the expression, so the toml would look like:

```toml
kernel_0 = '/kernels/kernel_0'  # filepath to 1st serialized kernel
kernel_1 = '/kernels/kernel_1'  # filepath to 2nd serialized kernel
kernel = '(kernel_0 + kernel_1)/2'  # expression defining combined kernel
```

The second option is to simply serialize the entire combined kernel. Hopefully this will be simple if `SumKernel` etc are `nn.Module` and `tf.keras.Model`. 

My preference is for the latter, since the former will likely get rather complex!